src/lib/ppx_coda/register_event.ml

```diff
@@ -0,0 +1,223 @@
+  (* check that every interpolation point $foo in msg has a matching label;
+     OK to have extra labels not mentioned in message
+  *)
+  let len = String.length msg in
+  let interpolates =
+    let re = Str.regexp "\\$[a-z]\\([a-z]\\|[0-9]\\)*" in
+    let rec loop start acc =
+      if start >= len then acc
+      else
+        try
+          let offs = Str.search_forward re msg start in
+          let s = Str.matched_string msg in
+          let s_len = String.length s in
+          loop (offs + s_len) (String.sub s ~pos:1 ~len:(s_len - 1) :: acc)
+        with _ ->
+          (* wild-card match; actual match is deprecated Not_found *)
+          acc
+    in
+    loop 0 []
+  in
+  List.iter interpolates ~f:(fun interp ->
+      if not (List.mem label_names interp ~equal:String.equal) then
+        Location.raise_errorf ~loc:msg_loc
+          (Scanf.format_from_string
+             (sprintf
+                "The msg contains interpolation point \"$%s\" which is not a \
+                 field in the record"
+                interp)
+             "") ) ;
+  let event_name = String.lowercase ctor in
+  let identifying = String.concat (event_name :: label_names) in
+  let id_string = hash identifying in
+  let (module Ast_builder) = Ast_builder.make ty_ext.ptyext_path.loc in
+  let open Ast_builder in
+  let core_type_of_string s =
+    ptyp_constr {txt= Longident.parse s; loc= Location.none} []
+  in
+  let id_name = event_name ^ "_id" in
+  let id_type = core_type_of_string (add_module_qualifier "id") in
+  let id_of_string = add_module_qualifier "id_of_string" in
+  let event_id =
+    [%stri
+      let ([%p pvar id_name] : [%t id_type]) =
+        [%e evar id_of_string] [%e estring id_string]]
+  in
+  let repr_name = event_name ^ "_repr" in
+  (* constructor declaration to build pattern *)
+  let ctor_decl =
+    { pcd_name= {txt= ctor; loc= Location.none}
+    ; pcd_args= record
+    ; pcd_res= None
+    ; pcd_loc= Location.none
+    ; pcd_attributes= [] }
+  in
+  let ctor_args =
+    let ids =
+      List.map label_names ~f:(fun label ->
+          ({txt= Lident label; loc= Location.none}, pvar label) )
+    in
+    { ppat_desc= Ppat_record (ids, Closed)
+    ; ppat_loc= Location.none
+    ; ppat_attributes= [] }
+  in
+  let json_pairs =
+    List.map label_decls ~f:(fun decl ->
+        let name = estring decl.pld_name.txt in
+        let var = evar decl.pld_name.txt in
+        let to_json_var = evar (to_yojson_of_core_type decl.pld_type) in
+        pexp_tuple [name; pexp_apply to_json_var [(Nolabel, var)]] )
+  in
+  let parse_args =
+    List.map label_names ~f:(fun label -> ppat_tuple [pstring label; pvar label]
+    )
+  in
+  let parse_args_pat =
+    let empty_list =
+      ppat_construct {txt= Lident "[]"; loc= Location.none} None
+    in
+    let cons = {txt= Lident "::"; loc= Location.none} in
+    List.fold_right parse_args ~init:empty_list ~f:(fun hd tl ->
+        ppat_construct cons (Some (ppat_tuple [hd; tl])) )
```

Comment:
Metaquot would make this terser and easier (for me) to follow.
```suggestion
    List.fold_right parse_args ~init:[%expr []] ~f:(fun hd tl -> [%expr [%e hd] :: [%e tl]])
```

Reply:
well, this is a pattern, not an expression, so it'd be somewhat different

Reply:
Fixed